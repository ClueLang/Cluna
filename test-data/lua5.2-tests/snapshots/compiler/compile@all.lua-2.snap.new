---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
