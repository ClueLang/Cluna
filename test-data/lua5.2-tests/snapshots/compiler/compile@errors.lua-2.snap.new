---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/errors.lua
---
print("testing errors");
local debug = require("debug")
local mt = getmetatable(_G) || {}
local oldmm = mt.__index
mt.__index = nil
global fn doit(s) {
    local f, msg = load(s)
    if f == nil {
        return msg
    }
    local cond, msg = pcall(f)
    return (!cond) && msg
}
global fn checkmessage(prog, msg) {
    local m = doit(prog)
    assert(string.find(m, msg, 1, true));
}
global fn checksyntax(prog, extra, token, line) {
    local msg = doit(prog)
    if !string.find(token, "^<%a") && !string.find(token, "^char%(") {
        token = "'" .. token .. "'"
    }
    token = string.gsub(token, "(%p)", "%%%1")
    local pt = string.format(`^%%[string ".*"%%]:%d: .- near %s$`, line, token)
    assert(string.find(msg, pt));
    assert(string.find(msg, msg, 1, true));
}
assert(doit("error('hi', 0)") == 'hi');
assert(doit("error()") == nil);
if !_no32 {
    assert(doit("table.unpack({}, 1, n=2^30)"));
}
assert(doit("a=math.sin()"));
assert(!doit("tostring(1)") && doit("tostring()"));
assert(doit("tonumber()"));
assert(doit("repeat until 1; a"));
assert(doit("return;;"));
assert(doit("assert(false)"));
assert(doit("assert(nil)"));
assert(doit("function a (... , ...) end"));
assert(doit("function a (, ...) end"));
assert(doit("local t={}; t = t[#t] + 1"));
checksyntax(`
  local a = {4

`, "'}' expected (to close '{' at line 1)", "<eof>", 3);
checkmessage("a=1; bbbb=2; a=math.sin(3)+bbbb(3)", "global 'bbbb'");
checkmessage("a=1; local a,bbbb=2,3; a = math.sin(1) and bbbb(3)", "local 'bbbb'");
checkmessage("a={}; do local a=1 end a:bbbb(3)", "method 'bbbb'");
checkmessage("local a={}; a.bbbb(3)", "field 'bbbb'");
assert(!string.find(doit("a={13}; local bbbb=1; a[bbbb](3)"), "'bbbb'"));
checkmessage("a={13}; local bbbb=1; a[bbbb](3)", "number");
checkmessage("a=(1)..{}", "a table value");
aaa = nil
checkmessage("aaa.bbb:ddd(9)", "global 'aaa'");
checkmessage("local aaa={bbb=1}; aaa.bbb:ddd(9)", "field 'bbb'");
checkmessage("local aaa={bbb={}}; aaa.bbb:ddd(9)", "method 'ddd'");
checkmessage("local a,b,c; (function () a = b+1 end)()", "upvalue 'b'");
assert(!doit("local aaa={bbb={ddd=next}}; aaa.bbb:ddd(nil)"));
checkmessage("local _ENV = {x={}}; a = a + 1", "global 'a'");
checkmessage("b=1; local aaa='a'; x=aaa+b", "local 'aaa'");
checkmessage("aaa={}; x=3/aaa", "global 'aaa'");
checkmessage("aaa='2'; b=nil;x=aaa*b", "global 'b'");
checkmessage("aaa={}; x=-aaa", "global 'aaa'");
assert(!string.find(doit("aaa={}; x=(aaa or aaa)+(aaa and aaa)"), "'aaa'"));
assert(!string.find(doit("aaa={}; (aaa or aaa)()"), "'aaa'"));
checkmessage("print(print < 10)", "function");
checkmessage("print(print < print)", "two function");
_G.D = debug
checkmessage(`
  -- create light udata
  local x = D.upvalueid(function () return debug end, 1)
  D.setuservalue(x, {})
`, "light userdata");
_G.D = nil
checkmessage("(io.write or print){}", "io.write");
checkmessage("(collectgarbage or print){}", "collectgarbage");
local t = {}
for i = 1, 1000 {
    t[(i)] = "a = x" .. i
}
local s = table.concat(t, "; ")
t = nil
checkmessage(s .. "; a = bbb + 1", "global 'bbb'");
checkmessage("local _ENV=_ENV;" .. s .. "; a = bbb + 1", "global 'bbb'");
checkmessage(s .. "; local t = {}; a = t.bbb + 1", "field 'bbb'");
checkmessage(s .. "; local t = {}; t:bbb()", "method 'bbb'");
checkmessage(`aaa=9
repeat until 3==3
local x=math.sin(math.cos(3))
if math.sin(1) == x then return math.sin(1) end   -- tail call
local a,b = 1, {
  {x='a'..'b'..'c', y='b', z=x},
  {1,2,3,4,5} or 3+3<=3+3,
  3+1>3+1,
  {d = x and aaa[x or y]}}
`, "global 'aaa'");
checkmessage(`
local x,y = {},1
if math.sin(1) == 0 then return 3 end    -- return
x.a()`, "field 'a'");
checkmessage(`
prefix = nil
insert = nil
while 1 do
  local a
  if nil then break end
  insert(prefix, a)
end`, "global 'insert'");
checkmessage(`  -- tail call
  return math.sin("a")
`, "'sin'");
checkmessage(`collectgarbage("nooption")`, "invalid option");
checkmessage(`x = print .. "a"`, "concatenate");
checkmessage("getmetatable(io.stdin).__gc()", "no value");
checkmessage(`
local Var
local function main()
  NoSuchName (function() Var=0 end)
end
main()
`, "global 'NoSuchName'");
print('+');
a = {}
setmetatable(a, {
    __index = string
});
checkmessage("a:sub()", "bad self");
checkmessage("string.sub('a', {})", "#2");
checkmessage("('a'):sub{}", "#1");
checkmessage("table.sort({1,2,3}, table.sort)", "'table.sort'");
checkmessage("string.gsub('s', 's', setmetatable)", "setmetatable'");
global fn f(n) {
    local c = coroutine.create(f)
    local a, b = coroutine.resume(c)
    return b
}
assert(string.find(f(), "C stack overflow"));
checkmessage("coroutine.yield()", "outside a coroutine");
f1 = fn () {
    table.sort({
        1, 
        2, 
        3
    }, coroutine.yield);
}
f = coroutine.wrap(fn () {
    return pcall(f1)
})
assert(string.find(select(2, f()), "yield across"));
idsize = 60 - 1
local fn checksize(source) {
    local _, msg = load("x", source)
    msg = string.match(msg, "^([^:]*):")
    assert(msg::len() <= idsize);
}
for i = 60 - 10, 60 + 10 {
    checksize("@" .. string.rep("x", i));
    checksize(string.rep("x", i - 10));
    checksize("=" .. string.rep("x", i));
}
local fn lineerror(s, l) {
    local err, msg = pcall(load(s))
    local line = string.match(msg, ":(%d+):")
    assert((line && line + 0) == l);
}
lineerror("local a\n for i=1,'a' do \n print(i) \n end", 2);
lineerror("\n local a \n for k,v in 3 \n do \n print(k) \n end", 3);
lineerror("\n\n for k,v in \n 3 \n do \n print(k) \n end", 4);
lineerror("function a.x.y ()\na=a+1\nend", 1);
lineerror("a = \na\n+\n{}", 3);
lineerror("a = \n3\n+\n(\n4\n/\nprint)", 6);
lineerror("a = \nprint\n+\n(\n4\n/\n7)", 3);
lineerror("a\n=\n-\n\nprint\n;", 3);
lineerror(`
a
(
23)
`, 1);
lineerror(`
local a = {x = 13}
a
.
x
(
23
)
`, 2);
lineerror(`
local a = {x = 13}
a
.
x
(
23 + a
)
`, 6);
local p = `
function g() f() end
function f(x) error('a', X) end
g()
`
X = 3
lineerror((p), 3);
X = 0
lineerror((p), nil);
X = 1
lineerror((p), 2);
X = 2
lineerror((p), 1);
if !_soft {
    C = 0
    local l = debug.getinfo(1, "l").currentline
    global fn y() {
        C = C + 1
        y();
    }
    local fn checkstackmessage(m) {
        return (string.find(m, "^.-:%d+: stack overflow"))
    }
    assert(checkstackmessage(doit('y()')));
    print('+');
    assert(checkstackmessage(doit('y()')));
    print('+');
    assert(checkstackmessage(doit('y()')));
    print('+');
    C = 0
    local l1
    local fn g(x) {
        l1 = debug.getinfo(x, "l").currentline
        y();
    }
    local _, stackmsg = xpcall(g, debug.traceback, 1)
    print('+');
    local stack = {}
    for line with string.gmatch(stackmsg, "[^\n]*") {
        local curr = string.match(line, ":(%d+):")
        if curr {
            table.insert(stack, tonumber(curr));
        }
    }
    local i = 1
    while stack[(i)] ~= l1 {
        assert(stack[(i)] == l);
        i = i + 1
    }
    assert(i > 15);
    local res, msg = xpcall(error, error)
    assert(!res && type(msg) == 'string');
    print('+');
    local fn f(x) {
        if x == 0 {
            error('a\n');
        } else {
            local aux = fn () {
                return f(x - 1)
            }
            local a, b = xpcall(aux, aux)
            return a, b
        }
    }
    f(3);
    local fn loop(x, y, z) {
        return 1 + loop(x, y, z)
    }
    local res, msg = xpcall(loop, fn (m) {
        assert(string.find(m, "stack overflow"));
        local res, msg = pcall(loop)
        assert(string.find(msg, "error handling"));
        assert(math.sin(0) == 0);
        return 15
    })
    assert(msg == 15);
    res, msg = pcall(fn () {
        for i = 999900, 1000000, 1 {
            table.unpack({}, 1, i);
        }
    })
    assert(string.find(msg, "too many results"));
}
global fn f() {
    error({
        msg = 'x'
    });
}
res, msg = xpcall(f, fn (r) {
    return {
        msg = r.msg .. 'y'
    }
})
assert(msg.msg == 'xy');
a, b, c = xpcall(string.find, error, "alo", "al")
assert(a && b == 1 && c == 2);
a, b, c = xpcall(string.find, fn (x) {
    return {}
}, true, "al")
assert(!a && type(b) == "table" && c == nil);
print('+');
checksyntax("syntax error", "", "error", 1);
checksyntax("1.000", "", "1.000", 1);
checksyntax("[[a]]", "", "[[a]]", 1);
checksyntax("'aa'", "", "'aa'", 1);
checksyntax("\255a = 1", "", "char(255)", 1);
doit('I = load("a=9+"); a=3');
assert(a == 3 && I == nil);
print('+');
lim = 1000
if _soft {
    lim = 100
}
for i = 1, lim {
    doit('a = ');
    doit('a = 4+nil');
}
local fn testrep(init, rep) {
    local s = "local a; " .. init .. string.rep(rep, 400)
    local a, b = load(s)
    assert(!a && string.find(b, "levels"));
}
testrep("a=", "{");
testrep("a=", "(");
testrep("", "a(");
testrep("", "do ");
testrep("", "while a do ");
testrep("", "if a then else ");
testrep("", "function foo () ");
testrep("a=", "a..");
testrep("a=", "a^");
local s = ("a,")::rep(200) .. "a=nil"
local a, b = load(s)
assert(!a && string.find(b, "levels"));
local lim = 127
local s = "local function fooA ()\n  local "
for j = 1, lim {
    s = s .. "a" .. j .. ", "
}
s = s .. "b,c\n"
s = s .. "local function fooB ()\n  local "
for j = 1, lim {
    s = s .. "b" .. j .. ", "
}
s = s .. "b\n"
s = s .. "function fooC () return b+c"
local c = 1 + 2
for j = 1, lim {
    s = s .. "+a" .. j .. "+b" .. j
    c = c + 2
}
s = s .. "\nend  end end"
local a, b = load(s)
assert(c > 255 && string.find(b, "too many upvalues") && string.find(b, "line 5"));
s = "\nfunction foo ()\n  local "
for j = 1, 300 {
    s = s .. "a" .. j .. ", "
}
s = s .. "b\n"
local a, b = load(s)
assert(string.find(b, "line 2"));
mt.__index = oldmm
print('OK');
