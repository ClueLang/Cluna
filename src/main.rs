#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
