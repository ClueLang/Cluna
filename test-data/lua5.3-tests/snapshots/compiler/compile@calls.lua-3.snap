---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.3-tests/calls.lua
---
print("testing functions and calls");
local debug = require("debug")
assert(type(1 < 2) == 'boolean');
assert(type(true) == 'boolean' && type(false) == 'boolean');
assert(type(nil) == 'nil' && type(-3) == 'number' && type('x') == 'string' && type({}) == 'table' && type(type) == 'function');
assert(type(assert) == type(print));
global fn f(x) {
    return a::x(x)
}
assert(type(f) == 'function');
assert(notpcall(type));
{
    local tostring = _ENV.tostring
    _ENV.tostring = nil
    local st, msg = pcall(print, 1)
    assert(st == false && string.find(msg, "attempt to call a nil value"));
    _ENV.tostring = fn () {
        return {}
    }
    local st, msg = pcall(print, 1)
    assert(st == false && string.find(msg, "must return a string"));
    _ENV.tostring = tostring
}
fact = false
{
    local res = 1
    local fn fact(n) {
        if n == 0 {
            return res
        } else {
            return n * fact(n - 1)
        }
    }
    assert(fact(5) == 120);
}
assert(fact == false);
a = {
    i = 10
}
self = 20
method a::x(x) {
    return x + self.i
}
method a.y(x) {
    return x + self
}
assert(a::x(1) + 10 == a.y(1));
a.t = {
    i = -100
}
a[("t")].x = fn (self, a, b) {
    return self.i + a + b
}
assert(a.t::x(2, 3) == -95);
{
    local a = {
        x = 0
    }
    method a::add(x) {
        self.x, a.y = self.x + x, 20
        return self
    }
    assert(a::add(10)::add(20)::add(30).x == 60 && a.y == 20);
}
local a = {
    b = {
        c = {}
    }
}
method a.b.c.f1(x) {
    return x + 1
}
method a.b.c::f2(x, y) {
    self[(x)] = y
}
assert(a.b.c.f1(4) == 5);
a.b.c::f2('k', 12);
assert(a.b.c.k == 12);
print('+');
t = nil
global fn f(a, b, c) {
    local d = 'a'
    t = {
        a, 
        b, 
        c, 
        d
    }
}
f(1, 2);
assert(t[(1)] == 1 && t[(2)] == 2 && t[(3)] == nil && t[(4)] == 'a');
f(1, 2, 3, 4);
assert(t[(1)] == 1 && t[(2)] == 2 && t[(3)] == 3 && t[(4)] == 'a');
global fn fat(x) {
    if x <= 1 {
        return 1
    } else {
        return x * load("return fat(" .. x - 1 .. ")", "")()
    }
}
assert(load("load 'assert(fat(6)==720)' () "))();
a = load('return fat(5), 3')
a, b = a()
assert(a == 120 && b == 3);
print('+');
global fn err_on_n(n) {
    if n == 0 {
        error();
        exit(1);
    } else {
        err_on_n(n - 1);
        exit(1);
    }
}
{
    global fn dummy(n) {
        if n > 0 {
            assert(notpcall(err_on_n, n));
            dummy(n - 1);
        }
    }
}
dummy(10);
global fn deep(n) {
    if n > 0 {
        deep(n - 1);
    }
}
deep(10);
deep(200);
global fn deep(n) {
    if n > 0 {
        return deep(n - 1)
    } else {
        return 101
    }
}
assert(deep(30000) == 101);
a = {}
method a::deep(n) {
    if n > 0 {
        return self::deep(n - 1)
    } else {
        return 101
    }
}
assert(a::deep(30000) == 101);
print('+');
a = nil(fn (x) {
    a = x
})(23)
assert(a == 23 && (fn (x) {
    return x * 2
})(20) == 40);
Z = fn (le) {
    local fn a(f) {
        return le(fn (x) {
            return f(f)(x)
        })
    }
    return a(a)
}
F = fn (f) {
    return fn (n) {
        if n == 0 {
            return 1
        } else {
            return n * f(n - 1)
        }
    }
}
fat = Z(F)
assert(fat(0) == 1 && fat(4) == 24 && Z(F)(5) == 5 * Z(F)(4));
local fn g(z) {
    local fn f(a, b, c, d) {
        return fn (x, y) {
            return a + b + c + d + a + x + y + z
        }
    }
    return f(z, z + 1, z + 2, z + 3)
}
f = g(10)
assert(f(9, 16) == 10 + 11 + 12 + 13 + 10 + 9 + 16 + 10);
Z, F, f = nil
print('+');
global fn unlpack(t, i) {
    i = i || 1
    if (i <= #t) {
        return t[(i)], unlpack(t, i + 1)
    }
}
global fn equaltab(t1, t2) {
    assert(#t1 == #t2);
    for i = 1, #t1 {
        assert(t1[(i)] == t2[(i)]);
    }
}
local pack = fn (...) {
    return (table.pack(...))
}
global fn f() {
    return 1, 2, 30, 4
}
global fn ret2(a, b) {
    return a, b
}
local a, b, c, d = unlpack({
    1, 
    2, 
    3
})
assert(a == 1 && b == 2 && c == 3 && d == nil);
a = {
    1, 
    2, 
    3, 
    4, 
    false, 
    10, 
    'alo', 
    false, 
    assert
}
equaltab(pack(unlpack(a)), a);
equaltab(pack(unlpack(a), -1), {
    1, 
    -1
});
a, b, c, d = ret2(f()), ret2(f())
assert(a == 1 && b == 1 && c == 2 && d == nil);
a, b, c, d = unlpack(pack(ret2(f()), ret2(f())))
assert(a == 1 && b == 1 && c == 2 && d == nil);
a, b, c, d = unlpack(pack(ret2(f()), (ret2(f()))))
assert(a == 1 && b == 1 && c == nil && d == nil);
a = ret2({
    unlpack({
        1, 
        2, 
        3
    }), 
    unlpack({
        3, 
        2, 
        1
    }), 
    unlpack({
        "a", 
        "b"
    })
})
assert(a[(1)] == 1 && a[(2)] == 3 && a[(3)] == "a" && a[(4)] == "b");
rawget({}, "x", 1);
rawset({}, "x", 1, 2);
assert(math.sin(1, 2) == math.sin(1));
table.sort({
    10, 
    9, 
    8, 
    4, 
    19, 
    23, 
    0, 
    0
}, fn (a, b) {
    return a < b
}, "extra arg");
local x = "-- a comment\0\0\0\n  x = 10 + \n23; \
     local a = function () x = 'hi' end; \
     return '\0'"
global fn read1(x) {
    local i = 0
    return fn () {
        collectgarbage();
        i = i + 1
        return string.sub(x, i, i)
    }
}
global fn cannotload(msg, a, b) {
    assert(nota && string.find(b, msg));
}
a = assert(load(read1(x), "modname", "t", _G))
assert(a() == "\0" && _G.x == 33);
assert(debug.getinfo(a).source == "modname");
cannotload("attempt to load a text chunk", load(read1(x), "modname", "b", {}));
cannotload("attempt to load a text chunk", load(x, "modname", "b"));
a = assert(load(fn () {
    return nil
}))
a();
assert(notload(fn () {
    return true
}));
local t = {
    nil, 
    "return ", 
    "3"
}
f, msg = load(fn () {
    return table.remove(t, 1)
})
assert(f() == nil);
f = load(string.dump(fn () {
    return 1
}), nil, "b", {})
assert(type(f) == "function" && f() == 1);
x = string.dump(load("x = 1; return x"))
a = assert(load(read1(x), nil, "b"))
assert(a() == 1 && _G.x == 1);
cannotload("attempt to load a binary chunk", load(read1(x), nil, "t"));
cannotload("attempt to load a binary chunk", load(x, nil, "t"));
assert(notpcall(string.dump, print));
cannotload("unexpected symbol", load(read1("*a = 123")));
cannotload("unexpected symbol", load("*a = 123"));
cannotload("hhi", load(fn () {
    error("hhi");
}));
assert(load("return _ENV", nil, nil, 123)() == 123);
local x
XX = 123
local fn h() {
    local y = x
    return XX
}
local d = string.dump(h)
x = load(d, "", "b")
assert(debug.getupvalue(x, 2) == '_ENV');
debug.setupvalue(x, 2, _G);
assert(x() == 123);
assert(assert(load("return XX + ...", nil, nil, {
    XX = 13
}))(4) == 17);
x = `
  return function (x)
    return function (y)
     return function (z)
       return x+y+z
     end
   end
  end
`
a = assert(load(read1(x)))
assert(a()(2)(3)(10) == 15);
local a, b = 20, 30
x = load(string.dump(fn (x) {
    if x == "set" {
        a = 10 + b
        b = b + 1
    } else {
        return a
    }
}), "", "b", nil)
assert(x() == nil);
assert(debug.setupvalue(x, 1, "hi") == "a");
assert(x() == "hi");
assert(debug.setupvalue(x, 2, 13) == "b");
assert(notdebug.setupvalue(x, 3, 10));
x("set");
assert(x() == 23);
x("set");
assert(x() == 24);
{
    local nup = 200
    local prog = {
        "local a1"
    }
    for i = 2, nup {
        prog[(#prog + 1)] = ", a" .. i
    }
    prog[(#prog + 1)] = " = 1"
    for i = 2, nup {
        prog[(#prog + 1)] = ", " .. i
    }
    local sum = 1
    prog[(#prog + 1)] = "; return function () return a1"
    for i = 2, nup {
        prog[(#prog + 1)] = " + a" .. i
        sum = sum + i
    }
    prog[(#prog + 1)] = " end"
    prog = table.concat(prog)
    local f = assert(load(prog))()
    assert(f() == sum);
    f = load(string.dump(f))
    local a = 10
    local h = fn () {
        return a
    }
    for i = 1, nup {
        debug.upvaluejoin(f, i, h, 1);
    }
    assert(f() == 10 * nup);
}
{
    local t = {
        x = 1
    }
    method t::_012345678901234567890123456789012345678901234567890123456789() {
        return self.x
    }
    assert(t::_012345678901234567890123456789012345678901234567890123456789() == 1);
}
assert((fn () {
    return nil
})(4) == nil);
assert((fn () {
    local a
    return a
})(4) == nil);
assert((fn (a) {
    return a
})() == nil);
print("testing binary chunks");
{
    local header = string.pack("c4BBc6BBBBBj", "\27Lua", 5 * 16 + 3, 0, "\x19\x93\r\n\x1a\n", string.packsize("i"), string.packsize("T"), 4, string.packsize("j"), string.packsize("n"), 0x5678)
    local c = string.dump(fn () {
        local a = 1
        local b = 3
        return a + b * 3
    })
    assert(string.sub(c, 1, #header) == header);
    for i = 1, #header {
        local s = string.sub(c, 1, i - 1) .. string.char(string.byte(string.sub(c, i, i)) + 1) .. string.sub(c, i + 1, -1)
        assert(#s == #c);
        assert(notload(s));
    }
    for i = 1, #c - 1 {
        local st, msg = load(string.sub(c, 1, i))
        assert(notst && string.find(msg, "truncated"));
    }
    assert(assert(load(c))() == 10);
}
print('OK');
return deep
